src/mlpack/methods/matrix_completion/matrix_completion.hpp

```diff
@@ -13,50 +14,65 @@
 #ifndef MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_HPP
 #define MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_HPP
 
-#include <mlpack/core/optimizers/sdp/sdp.hpp>
-#include <mlpack/core/optimizers/sdp/lrsdp.hpp>
+#include "mc_sdp_solver.hpp"
+#include "mc_fw_solver.hpp"
 
 namespace mlpack {
 namespace matrix_completion {
 
 /**
- * This class implements the popular nuclear norm minimization heuristic for
- * matrix completion problems. That is, given known values M_ij's, the
- * following optimization problem (semi-definite program) is solved to fill in
- * the remaining unknown values of X
+ * This class implements a wrapper for different types of solvers of matrix
+ * completion problems. That is, we want to recover a large low rank matrix,
+ * with only a few elements in the matrix known. There are quite many
+ * mathematical models and numerical algorithms to solve this kind of problems.
+ * This class only provide a wrapper using class templates.
  *
- *   min ||X||_* subj to X_ij = M_ij
+ * The model solved using SDP solver is:
+ *
+ * \f[
+ *   min ||X||_* subj to X_ij = M_ij, \qquad (i,j) \in \Omega
+ * \f]
  *
  * where ||X||_* denotes the nuclear norm (sum of singular values of X).
  *
- * For a theoretical treatment of the conditions necessary for exact recovery,
- * see the following paper:
+ * For a theoretical treatment of the conditions necessary for exact recovery
+ * using the above model, see the following paper:
  *
  *   A Simpler Appoarch to Matrix Completion.
  *   Benjamin Recht. JMLR 11.
  *   http://arxiv.org/pdf/0910.0651v2.pdf
  *
- * An example of how to use this class is shown below:
+ * The model solved using FrankWolfe solver is:
+ *
+ * \f[
+ * min \sum_{(i,j) \in \Omega} 0.5 * (X_ij - M_ij)^2, \qquad
+ * s.t. ||X||_* <= tau
+ * \f]
+ *
+ *
+ * Example of the usage of the code is:
  *
  * @code
  * size_t m, n;         // size of unknown matrix
  * arma::umat indices;  // contains the known indices [2 x n_entries]
  * arma::vec values;    // contains the known values [n_entries]
  * arma::mat recovered; // will contain the completed matrix
  *
- * MatrixCompletion mc(m, n, indices, values);
+ * MatrixCompletion<MCSolverType> mc(m, n, indices, values);
  * mc.Recover(recovered);
  * @endcode
  *
- * @see LRSDP
  */
+template<typename MCSolverType>
```

Comment:
I like templatizing the solver, but I think there are a couple of ways we can improve it here.  From what I can tell, the two solvers you are considering---`MCSDPSolver` and `MCFWSolver`---have different APIs, and this means that half of the functions of `MatrixCompletion` can only be used with one solver and half can only be used with the other.  We should probably find a way to instead change the APIs of each solver to match, so that the templatization can work with either and the API is unified.

From what I can tell, the primary problem seems to be that `MCFWSolver()` accepts a parameter `tau` whereas `MCSDPSolver()` accepts either `initialPoint` or `r` (the maximum rank).  There is no clear relation (I think) between `tau` and `r`, so we can't combine the parameters.  Instead, I might suggest the following pattern which is used in other places in mlpack:

```
// Construct the MatrixCompletion object with the given optimizer.
MatrixCompletion(
    const size_t m,
    const size_t n,
    const arma::umat& indices,
    const arma::vec& values,
    MCSolverType solver = MCSolverType()); // use std::move(solver) to avoid copying the solver in the implementation

// Access the solver to get its parameters.
const MCSolverType& Solver() const { return solver; }
// Modify the solver.
MCSolverType& Solver() { return solver; }
```

Internally, I would suggest having `MatrixCompletion` hold the `MatrixCompletionFWFunction` or `LRSDP<SDP<arma::sp_mat>>` objects.  You could expose a typedef in `MCSDPSolver` and `MCFWSolver` and call it something like `FunctionType`; then inside of `MatrixCompletion` you can hold `FunctionType function`.  Separating the function from the optimizer becomes important for the second usage example below.

Once all this is done, a user can set their given tau in one of two ways:

```
MatrixCompletion<MCFWSolver> m(...);
m.Solver().Tau() = myFavoriteTauValue;
// or alternately
MCFWSolver solver(myFavoriteTauValue); // all function-related parameters should move out of the solver
MatrixCompletion<MCFWSolver> m(..., std::move(solver));
```

I hope what I've written here makes sense.  Basically, the key is that I think we can restructure the design of the solver and function classes so that they all fit nicely into `MatrixCompletion`.  Let me know if you agree, or if I can clarify what I mean.

Reply:
Thanks a lot for this idea! I was thinking that the way I did is really not elegant. I will change the structure to your suggestion.
I think I basically understand what you are suggesting. I just need a constructor of each `MCSolverType` with no input, and different parameters could be set later. Right?

As to the `FunctionType`, I just have a C++ question:
If `FunctionType` is a typedef defined in each solver, and the solver is a member in `MatrixCompletion` class. Can `MatrixCompletion` class know how `FunctionType` is defined?

Reply:
Great, I think the new structure will be an improvement.  Thanks!

`MatrixCompletion` can know what `FunctionType` is by using `SolverType::FunctionType`.  So, e.g., if I have a simple set of code like...

```
class MySolver
{
 public:
  typedef MyFunction FunctionType;
};

template<typename SolverType>
class MatrixCompletion
{
  typename SolverType::FunctionType function;
};
```

then if I create a `MatrixCompletion<MySolver>` type, then its `function` value will have type `MyFunction`.  I hope I understood the question correctly; let me know if not.